mnist_forward_forward/main.py

```diff
@@ -0,0 +1,183 @@
+                )
+            ).mean()
+            self.opt.zero_grad()
+            loss.backward()
+            self.opt.step()
+            if i % args.log_interval == 0:
+                print("Loss: ", loss.item())
+        return self.forward(x_pos).detach(), self.forward(x_neg).detach()
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--epochs",
+        type=int,
+        default=1000,
+        metavar="N",
+        help="number of epochs to train (default: 10)",
+    )
+    parser.add_argument(
+        "--lr",
+        type=float,
+        default=0.03,
+        metavar="LR",
+        help="learning rate (default: 0.03)",
+    )
+    parser.add_argument(
+        "--no_cuda", action="store_true", default=False, help="disables CUDA training"
+    )
+    parser.add_argument(
+        "--no_mps", action="store_true", default=False, help="disables MPS training"
+    )
+    parser.add_argument(
+        "--seed", type=int, default=1, metavar="S", help="random seed (default: 1)"
+    )
+    parser.add_argument(
+        "--save_model",
+        action="store_true",
+        default=False,
+        help="For saving the current Model",
+    )
+    parser.add_argument(
+        "--train_size", type=int, default=50000, help="size of training set"
+    )
+    parser.add_argument(
+        "--threshold", type=float, default=2, help="threshold for training"
+    )
+    parser.add_argument("--test_size", type=int, default=10000, help="size of test set")
+    parser.add_argument(
+        "--save-model",
+        action="store_true",
+        default=False,
+        help="For Saving the current Model",
+    )
+    parser.add_argument(
+        "--log-interval",
+        type=int,
+        default=10,
+        metavar="N",
+        help="how many batches to wait before logging training status",
+    )
+    args = parser.parse_args()
+    use_cuda = not args.no_cuda and torch.cuda.is_available()
+    use_mps = not args.no_mps and torch.backends.mps.is_available()
+    if use_cuda:
+        device = torch.device("cuda")
+    elif use_mps:
+        device = torch.device("mps")
+    else:
+        device = torch.device("cpu")
+
+    train_kwargs = {"batch_size": args.train_size}
+    test_kwargs = {"batch_size": args.test_size}
+
+    if use_cuda:
+        cuda_kwargs = {"num_workers": 1, "pin_memory": True, "shuffle": True}
+        train_kwargs.update(cuda_kwargs)
+        test_kwargs.update(cuda_kwargs)
+
+    transform = Compose(
+        [
+            ToTensor(),
+            Normalize((0.1307,), (0.3081,)),
+            Lambda(lambda x: torch.flatten(x)),
+        ]
+    )
+    train_loader = DataLoader(
+        MNIST("./data/", train=True, download=True, transform=transform), **train_kwargs
+    )
+    test_loader = DataLoader(
+        MNIST("./data/", train=False, download=True, transform=transform), **test_kwargs
+    )
+    net = Net([784, 500, 500])
```

Comment:
Well, as you stated, @msaroufim, the algorithm does not converge on a single epoch because it assigns a single label to every instance. As a result, the train and test errors are approximately 90% accurate all the time (for example, as you may be aware with MNIST if you assign every x a single label, you are correct 90% of the time, always).

Because of this, the train and test errors for a single epoch are both about 90.

however,
Default epochs, are 1000, and mistakes are:

test error: 0.07350003719329834 
train error: 0.07302004098892212